tsfc/coffee_mode.py

```diff
@@ -0,0 +1,317 @@
+from gem.optimise import (replace_division, associate_sum, associate_product,
+                          unroll_indexsum, replace_delta, remove_componenttensors)
+from gem.refactorise import (MonomialSum, ATOMIC, COMPOUND, OTHER,
+                             collect_monomials)
+from gem.node import traversal
+from gem.gem import (Product, Sum, Comparison, Conditional, Division, Indexed,
+                     IndexSum, MathFunction, Power, Failure, one, index_sum,
+                     Terminal, ListTensor, FlexiblyIndexed, LogicalAnd,
+                     LogicalNot, LogicalOr)
+
+
+import tsfc.vanilla as vanilla
+
+flatten = vanilla.flatten
+
+finalise_options = dict(replace_delta=False, remove_componenttensors=False)
+
+
+def Integrals(expressions, quadrature_multiindex, argument_multiindices, parameters):
+    """Constructs an integral representation for each GEM integrand
+    expression.
+
+    :arg expressions: integrand multiplied with quadrature weight;
+                      multi-root GEM expression DAG
+    :arg quadrature_multiindex: quadrature multiindex (tuple)
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argument
+    :arg parameters: parameters dictionary
+
+    :returns: list of integral representations
+    """
+    # Unroll
+    max_extent = parameters["unroll_indexsum"]
+    if max_extent:
+        def predicate(index):
+            return index.extent <= max_extent
+        expressions = unroll_indexsum(expressions, predicate=predicate)
+    # Choose GEM expression as the integral representation
+    expressions = [index_sum(e, quadrature_multiindex) for e in expressions]
+    expressions = replace_delta(expressions)
+    expressions = remove_componenttensors(expressions)
+    expressions = replace_division(expressions)
+    return optimise_expressions(expressions, quadrature_multiindex, argument_multiindices)
+
+
+def optimise_expressions(expressions, quadrature_multiindex, argument_multiindices):
+    """Perform loop optimisations on GEM DAGs
+
+    :arg expressions: list of GEM DAGs
+    :arg quadrature_multiindex: quadrature multiindex (tuple)
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argumen
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+
+    def classify(argument_indices, expression):
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, (Indexed, Conditional)):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+
+    argument_indices = tuple([i for indices in argument_multiindices for i in indices])
```

Comment:
Amended. (I use `itertools.chain` and `itertools.count` as the names are likely to clash with variables)